File: 1-500/0236. Lowest Common Ancestor of a Binary Tree.java
// Medium
// DFS
// Time:O(n), Space:O(h)
// https://leetcode.cn/problems/lowest-common-ancestor-of-a-binary-tree/

class Solution {
    // 1. p in left & q in right 2.p in right & q in left 3. node itself
    // Start from root and recursively check if left and right has p or q
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if (root == null || root == p || root == q) return root;
        TreeNode left = lowestCommonAncestor(root.left, p, q);
        TreeNode right = lowestCommonAncestor(root.right, p, q);
        if (left == null && right == null) {
            return null;
        } else if (left != null && right == null) {
            return left;
        } else if (left == null && right != null) {
            return right;
        } else {
            return root;
        }
    }
}

File: 0001-0500/0236. Lowest Common Ancestor of a Binary Tree.go
package main
// Medium
// Time:O(n), Space:O(h)
// https://leetcode.cn/problems/lowest-common-ancestor-of-a-binary-tree/

func lowestCommonAncestor(root, p, q *TreeNode) *TreeNode {
    if root == nil || root == p || root == q {
        return root
    }

    left := lowestCommonAncestor(root.Left, p, q)
    right := lowestCommonAncestor(root.Right, p, q)

    if left != nil && right != nil {
        return root
    }

    if left != nil {
        return left
    }
    return right
}
